internal/query/cache.go

```diff
@@ -0,0 +1,82 @@
+package query
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/zitadel/logging"
+
+	"github.com/zitadel/zitadel/internal/cache"
+	"github.com/zitadel/zitadel/internal/cache/gomap"
+	"github.com/zitadel/zitadel/internal/cache/noop"
+)
+
+type Caches struct {
+	connectors *cacheConnectors
+	instance   cache.Cache[instanceIndex, string, *authzInstance]
```

Comment:
I defined connectors in anticipation of SQL and Redis backed caches. SO that the datebase connection can be defined once and then is reused for multiple caches. For example a single pgxpool shared between different object caches.

As you mentioned here: https://github.com/zitadel/zitadel/pull/8628#discussion_r1764500712 not all implementations will be able to support (or even need) auto pruning. For example in Redis, objects can be set with an expiry and Redis takes care of the pruning for you. On the flip side, in SQL we need to take care of pruning expired records. In the latter case the SQL connector's config will be a combination of connection parameters and the AutoPruneConfig. The in-memory cache doesn't need any connection parameters and only the AutoPruneConfig remains.

Once a cache uses a connector that has AutoPrune enabled, it will also start the pruner for its indices.

At the moment we have 1 cache and 1 connector, so indeed it looks a bit awkward now.

Reply:
great that answers my question thanks